### PR TITLE
Texture filtering support, font rendering and some fixes

### PR DESCRIPTION
**Pull request by digetx**
Add texture filtering support to grate. The font rendering is used by a new texture-filter grate test.

### CODE REVIEW

File: src/libgrate/fragment_asm_parser.y [hunk 1]
```diff
@@ -765,6 +767,19 @@ TEX_INSTRUCTION:
...
+TEX_LOD_REG:
+	',' T_ROW_REGISTER
+	{
+		asm_tex_instructions[asm_fs_instructions_nb].enable_bias = 1;
+		yyval.u = $2;
+	}
+	|
+	{
+		asm_tex_instructions[asm_fs_instructions_nb].enable_bias = 0;
+		yyval.u = 0xDEAD;
```
Comment (kusma) [inline src/libgrate/fragment_asm_parser.y]: The "common" way of doing this, is to consider these two different opcodes, [TEX](http://gallium.readthedocs.io/en/latest/tgsi.html#opcode-TEX) and [TXB](http://gallium.readthedocs.io/en/latest/tgsi.html#opcode-TXB)...

Comment (digetx) [reply src/libgrate/fragment_asm_parser.y]: Indeed, differentiating instructions by opcode should be more expressive and there will be more opcodes in the future. We should probably omit the 3-rd cubemap coord source register from the current TEX instruction and use it in the TEX2 instruction opcode (following TGSI ISA) once will add support for cubemaps. 

 I'll re-spin the series.

Comment (kusma) [reply src/libgrate/fragment_asm_parser.y]: TEX2 is about shadow cube-map arrays, though, how is that related?

Comment (digetx) [reply src/libgrate/fragment_asm_parser.y]: Ah, okay :) I just quickly glanced through the TGSI ISA without paying much attention. So the TEX opcode is fine as it is now, no need to change anything about it.

**Comment by digetx** [comment]
Re-spun, now with the TXB opcode. Please take a look =)

File: src/libgrate/grate-texture.c [hunk 2]
```diff
@@ -184,3 +264,212 @@ void grate_texture_clear(struct grate *grate, struct grate_texture *tex,
...
+		return 4;
+	case 5 ... 8:
+		return 3;
+	case 3 ... 4:
+		return 2;
+	case 2:
+		return 1;
+	case 0 ... 1:
+		return 0;
+	}
```
Comment (kusma) [inline src/libgrate/grate-texture.c]: This is a huge jump-table, using compiler-specific range-based case-statements... How about just doing __builtin_clz() instead? :)

File: src/libgrate/grate-3d.c [hunk 3]
```diff
@@ -659,11 +659,13 @@ static void grate_3d_set_texture_desc(struct host1x_pushbuf *pb,
 				      struct host1x_pixelbuffer *pixbuf,
 				      unsigned max_lod,
 				      unsigned wrap_mode,
-				      bool mip_filter,
-				      bool mag_filter,
-				      bool min_filter)
+				      bool mipmap_enabled,
+				      bool min_filter_linear_within,
+				      bool min_filter_linear_between,
+				      bool mag_filter_linear)
```
Comment (kusma) [inline src/libgrate/grate-3d.c]: This isn't really important, but I'm having a bit of difficulty liking this name-change. I know that I merged it in envytools, but I'm having a bit of second thought here... What's wrong with `mag`, `min`, `mip` names? These are the names that D3D uses for the filtering modes, I've never seen OpenGL refer to these apart from that wiki-site you listed (which is more of a textual explanation than "terminology")...

Basically, every time I read these, I get a little bit confused. I guess I'm just so used to the old names...

File: src/libgrate/grate-texture.c (hunk 2 above)
Comment (digetx) [reply src/libgrate/grate-texture.c]: That's not true. I checked disassembly and the generated code is tiny, actually it is compiled into

if (size > 1024) {
} else if (size > 512) {
} ...

File: src/libgrate/grate-3d.c (hunk 3 above)
Comment (digetx) [reply src/libgrate/grate-3d.c]: I'm not used to any naming convention, so do not really care. Just tell me how to name them.

MIPFILTER == MINFILTER_LINEAR_WITHIN ?
MINFILTER == MINFILTER_LINEAR_BETWEEN ?

Is that ^ correct?

Comment (kusma) [reply src/libgrate/grate-3d.c]: The "min"-filter is what happens to a single texture image when it's being minified (similar to magfilter), and the "mip"-filter is how the results of the min/mag-filters gets combined. So unless I'm confused, I think that would be:

MINFILTER == MINFILTER_LINEAR_WITHIN
MIPFILTER == MINFILTER_LINEAR_BETWEEN

Comment (digetx) [reply src/libgrate/grate-3d.c]: Okay, I'll switch to these names. Let me know if there is anything else confusing :)

File: src/libgrate/grate-texture.c (hunk 2 above)
Comment (kusma) [reply src/libgrate/grate-texture.c]: Oh, OK. I still think a single instruction is slightly better, but it's not at all important. Let's just leave it like this for now.

File: src/libgrate/grate-3d.c (hunk 3 above)
Comment (kusma) [reply src/libgrate/grate-3d.c]: @digetx: Nah, I think it all looks very reasonable. Sorry for being a bit nit-picky here ;)

File: src/libgrate/grate-texture.c (hunk 2 above)
Comment (digetx) [reply src/libgrate/grate-texture.c]: clz would also work, but I like more expressive approach. It would be insane if any relatively modern compiler won't optimize that switch-case. And again, the worst performance bottle neck is memory management, BO allocation time in particular.

File: src/libgrate/grate-3d.c (hunk 3 above)
Comment (digetx) [reply src/libgrate/grate-3d.c]: Since I'm not very familiar with OpenGL and related terminology in general, I'm expecting that from you :) Don't shy, always let me know if something feels clumsy.

File: src/libgrate/grate-texture.c (hunk 2 above)
Comment (kusma) [reply src/libgrate/grate-texture.c]: Fair enough!

Comment (digetx) [reply src/libgrate/grate-texture.c]: Yeah, one obvious optimization would be getting rid of CMA allocation on each command submit in the kernel driver. That gave a noticeable boost in performance while I was tackling the opentegra driver, actually not so long time ago. Also the current in-kernel command stream validation code completely kills performance, I usually have it turned off while hacking-developing.

**Comment by digetx** [comment]
Changed the names, please take a look. 🔍 :)

File: src/libgrate/grate-texture.c [hunk 4]
```diff
@@ -184,3 +268,216 @@ void grate_texture_clear(struct grate *grate, struct grate_texture *tex,
 	if (err < 0)
 		grate_error("host1x_gr2d_clear() failed: %d\n", err);
 }
+
+static unsigned log2_size(unsigned size)
```
Comment (kusma) [inline src/libgrate/grate-texture.c]: We already have a function in grate-3d.c called `log2_size` that has slightly different semantics. I find this a bit confusing, perhaps we could use a different name for one of the, or use this version for both and check `log2_width == pixbuf->width && log2_height == pixbug->height` instead of `log2_width >= 0 && log2_height >= 0` in `grate_3d_set_texture_desc`?

**Comment by digetx** [comment]
Consolidated log2_size() and switched to clz.

File: src/libgrate/grate-3d.h [hunk 5]
```diff
@@ -3,6 +3,8 @@
 
 #include <stdint.h>
 
+#define log2_size(s)		(31 - __builtin_clz(s))
```
Comment (digetx) [inline src/libgrate/grate-3d.h:6]: Here it is.

**Review by kusma — APPROVED** [review]
Nice!

**Comment by digetx** [comment]
You are too fast! I was going to write that I shouldn't have removed the grate_texture_set_max_lod() :)

Will re-add it.

**Comment by kusma** [comment]
I didn't pull yet!

**Comment by digetx** [comment]
Ah nice :) Will fix now then.

**Comment by digetx** [comment]
Factored out grate_create_texture2() addition to a standalone commit and added test4_max_lod() to texture-filter test. Should be good now :)

**Comment by kusma** [comment]
Great stuff :)
